<a href="https://colab.research.google.com/github/zairapg/hiperblog/blob/main/Actividad_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# Aprendizaje Supervisado

   Zaira Pizo Gurrute.

## Etapa 1: Entendimiento de los datos
Antes de comenzar incluso a preprocesar los datos, debemos conocer las características del conjunto de datos que tenemos.

Como mínimo debemos saber:
-  Cuántos registros hay?
-  Cuantos atributos tiene cada registro?
-  De qué tipo son los atributos?

Además:
-  Obtener una medida de centralidad para cada atributo
-  Obtener una medida de dispersión o desviación para los atributos que ésto tenga sentido
-  Calcular la matriz de correlación de los atributos de entrada para identificar atributos redundantes

Como paso previo al preprocesamiento de los datos, debemos:
-  Detectar si hay datos faltantes (determinar en qué columnas y cuantos datos faltan en un mismo registro)
-  Detectar si hay datos atípicos
-  Detectar si hay desbalance entre clases

Notar que en esta etapa no hacemos ninguna modificación sobre los datos, sólo estamos conociéndolos a fondo para saber con qué material contamos para trabajar.


In [4]:
!pip install ucimlrepo

In [71]:
# Carga de librerías y lectura del archivo que contiene los datos

import numpy as np
import pandas as pd
import sklearn as sk
import seaborn as sns
import openpyxl as xl
import matplotlib.pyplot as plt
from ucimlrepo import fetch_ucirepo

# 1. Se realiza el cargue del conjunto de datos en un data frame de pandas (data).


# fetch dataset
breast_cancer_wisconsin_original = fetch_ucirepo(id=15)

# data (as pandas dataframes)
X = breast_cancer_wisconsin_original.data.features
y = breast_cancer_wisconsin_original.data.targets

# metadata
print(breast_cancer_wisconsin_original.metadata)

# variable information
print(breast_cancer_wisconsin_original.variables)

data = X.copy()
data['target'] = y

# Imprimir los primeros registros para verificar
print(data.head())

data.to_excel('D:\datos_copy.xlsx', index=False)





{'uci_id': 15, 'name': 'Breast Cancer Wisconsin (Original)', 'repository_url': 'https://archive.ics.uci.edu/dataset/15/breast+cancer+wisconsin+original', 'data_url': 'https://archive.ics.uci.edu/static/public/15/data.csv', 'abstract': 'Original Wisconsin Breast Cancer Database', 'area': 'Health and Medicine', 'tasks': ['Classification'], 'characteristics': ['Multivariate'], 'num_instances': 699, 'num_features': 9, 'feature_types': ['Integer'], 'demographics': [], 'target_col': ['Class'], 'index_col': ['Sample_code_number'], 'has_missing_values': 'yes', 'missing_values_symbol': 'NaN', 'year_of_dataset_creation': 1990, 'last_updated': 'Sun Mar 10 2024', 'dataset_doi': '10.24432/C5HP4Z', 'creators': ['WIlliam Wolberg'], 'intro_paper': None, 'additional_info': {'summary': "Samples arrive periodically as Dr. Wolberg reports his clinical cases. The database therefore reflects this chronological grouping of the data. This grouping information appears immediately below, having been removed fro

In [38]:
#Cuál es el número de registros? #Cuál es el número de atributos?

shape = data.shape
shape

(699, 10)

In [39]:
#Cual es el tipo de los atributos?
data.dtypes

Clump_thickness                  int64
Uniformity_of_cell_size          int64
Uniformity_of_cell_shape         int64
Marginal_adhesion                int64
Single_epithelial_cell_size      int64
Bare_nuclei                    float64
Bland_chromatin                  int64
Normal_nucleoli                  int64
Mitoses                          int64
target                           int64
dtype: object

In [40]:
#Medida de centralidad y desviación para atributos numéricos:
data.describe()

Clump_thickness  Uniformity_of_cell_size  Uniformity_of_cell_shape  \
count       699.000000               699.000000                699.000000   
mean          4.417740                 3.134478                  3.207439   
std           2.815741                 3.051459                  2.971913   
min           1.000000                 1.000000                  1.000000   
25%           2.000000                 1.000000                  1.000000   
50%           4.000000                 1.000000                  1.000000   
75%           6.000000                 5.000000                  5.000000   
max          10.000000                10.000000                 10.000000   

       Marginal_adhesion  Single_epithelial_cell_size  Bare_nuclei  \
count         699.000000                   699.000000   683.000000   
mean            2.806867                     3.216023     3.544656   
std             2.855379                     2.214300     3.643857   
min             1.000000                     1.000000     1.000000   
25%             1.000000                     2.000000     1.000000   
50%             1.000000                     2.000000     1.000000   
75%             4.000000                     4.000000     6.000000   
max            10.000000                    10.000000    10.000000   

       Bland_chromatin  Normal_nucleoli     Mitoses      target  
count       699.000000       699.000000  699.000000  699.000000  
mean          3.437768         2.866953    1.589413    2.689557  
std           2.438364         3.053634    1.715078    0.951273  
min           1.000000         1.000000    1.000000    2.000000  
25%           2.000000         1.000000    1.000000    2.000000  
50%           3.000000         1.000000    1.000000    2.000000  
75%           5.000000         4.000000    1.000000    4.000000  
max          10.000000        10.000000   10.000000    4.000000

In [41]:
#Medida de centralidad para atributos categóricos:
data.mode()

Clump_thickness  Uniformity_of_cell_size  Uniformity_of_cell_shape  \
0                1                        1                         1   

   Marginal_adhesion  Single_epithelial_cell_size  Bare_nuclei  \
0                  1                            2          1.0   

   Bland_chromatin  Normal_nucleoli  Mitoses  target  
0                2                1        1       2

In [42]:
# Correlación entre los atributos de entrada numéricos
# Permite detectar si hay atributos redundantes (correlación mayor a 0.85 o menor a -0.85)
data.corr()

Clump_thickness  Uniformity_of_cell_size  \
Clump_thickness                     1.000000                 0.644913   
Uniformity_of_cell_size             0.644913                 1.000000   
Uniformity_of_cell_shape            0.654589                 0.906882   
Marginal_adhesion                   0.486356                 0.705582   
Single_epithelial_cell_size         0.521816                 0.751799   
Bare_nuclei                         0.593091                 0.691709   
Bland_chromatin                     0.558428                 0.755721   
Normal_nucleoli                     0.535835                 0.722865   
Mitoses                             0.350034                 0.458693   
target                              0.716001                 0.817904   

                             Uniformity_of_cell_shape  Marginal_adhesion  \
Clump_thickness                              0.654589           0.486356   
Uniformity_of_cell_size                      0.906882           0.705582   
Uniformity_of_cell_shape                     1.000000           0.683079   
Marginal_adhesion                            0.683079           1.000000   
Single_epithelial_cell_size                  0.719668           0.599599   
Bare_nuclei                                  0.713878           0.670648   
Bland_chromatin                              0.735948           0.666715   
Normal_nucleoli                              0.719446           0.603352   
Mitoses                                      0.438911           0.417633   
target                                       0.818934           0.696800   

                             Single_epithelial_cell_size  Bare_nuclei  \
Clump_thickness                                 0.521816     0.593091   
Uniformity_of_cell_size                         0.751799     0.691709   
Uniformity_of_cell_shape                        0.719668     0.713878   
Marginal_adhesion                               0.599599     0.670648   
Single_epithelial_cell_size                     1.000000     0.585716   
Bare_nuclei                                     0.585716     1.000000   
Bland_chromatin                                 0.616102     0.680615   
Normal_nucleoli                                 0.628881     0.584280   
Mitoses                                         0.479101     0.339210   
target                                          0.682785     0.822696   

                             Bland_chromatin  Normal_nucleoli   Mitoses  \
Clump_thickness                     0.558428         0.535835  0.350034   
Uniformity_of_cell_size             0.755721         0.722865  0.458693   
Uniformity_of_cell_shape            0.735948         0.719446  0.438911   
Marginal_adhesion                   0.666715         0.603352  0.417633   
Single_epithelial_cell_size         0.616102         0.628881  0.479101   
Bare_nuclei                         0.680615         0.584280  0.339210   
Bland_chromatin                     1.000000         0.665878  0.344169   
Normal_nucleoli                     0.665878         1.000000  0.428336   
Mitoses                             0.344169         0.428336  1.000000   
target                              0.756616         0.712244  0.423170   

                               target  
Clump_thickness              0.716001  
Uniformity_of_cell_size      0.817904  
Uniformity_of_cell_shape     0.818934  
Marginal_adhesion            0.696800  
Single_epithelial_cell_size  0.682785  
Bare_nuclei                  0.822696  
Bland_chromatin              0.756616  
Normal_nucleoli              0.712244  
Mitoses                      0.423170  
target                       1.000000

In [43]:
# Cual es el máximo de datos faltantes en un mismo registro?
# Si hay registros a los que les faltan muchos valores, es mejor eliminarlos.
max(data.isnull().sum(axis=1))

1

In [44]:
#Cuantos datos faltantes hay por cada atributo?
data.isnull().sum()

Clump_thickness                 0
Uniformity_of_cell_size         0
Uniformity_of_cell_shape        0
Marginal_adhesion               0
Single_epithelial_cell_size     0
Bare_nuclei                    16
Bland_chromatin                 0
Normal_nucleoli                 0
Mitoses                         0
target                          0
dtype: int64

In [48]:
#Cuantos registros hay por cada clase? es decir, por cada valor del atributo de salida?

print(data['target'].value_counts())

target
2    458
4    241
Name: count, dtype: int64


## Etapa 2: Preparación de los datos
De acuerdo a lo observado en la etapa 1, se define una secuencia de actividades que modifican los datos para eliminar las situaciones que puedan ser causa de fallo o deficiencia en el proceso de aprendizaje. En este caso se van a realizar las siguientes actividades:


1.   Completar los datos faltantes utilizando la media para atributos numéricos y la moda para atributos categóricos.
2.   Convertir los atributos categóricos a escala numérica.
3.   Normalizar los valores de algunos los atributos.
4.   Balancear las clases eliminando registros de la clase mayoritaria.






In [87]:
# Se imputa la mediana en la columna 'Bare_nuclei'
data['Bare_nuclei'] = data['Bare_nuclei'].fillna(data['Bare_nuclei'].median())
# Se verifica si aun hay valores faltantes
print(data.isnull().sum())

Clump_thickness                0
Uniformity_of_cell_size        0
Uniformity_of_cell_shape       0
Marginal_adhesion              0
Single_epithelial_cell_size    0
Bare_nuclei                    0
Bland_chromatin                0
Normal_nucleoli                0
Mitoses                        0
target                         0
dtype: int64


## Etapa 3: Modelado
En esta etapa se van a utilizar los datos para realizar aprendizaje del atributo target a partir de los valores de los demás atributos. Es decir, se va a predecir tumores como benignos o malignos. Las actividades a realizar son:
1.   Dividir el conjunto de datos en conjuntos de entrenamiento y test.
2.   Realizar el entrenamiento del modelo de regresión logística utilizando para ello el conjunto de datos de entrenamiento.
3.   Aplicar el modelo construido a los datos del conjunto de test.
4.   Cálculo de una métrica de desempeño para evaluar la calidad del modelo en sus prediciones sobre el conjunto de test.

In [89]:
# Se separan los datos en conjunto de entrenamiento y prueba.

from sklearn.model_selection import train_test_split

# Separar las características (X) y la variable objetivo (y)
X = data.drop(columns=['target'])
y = data['target']

# Dividir los datos en entrenamiento (80%) y prueba (20%)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Ver la forma de los conjuntos de entrenamiento y prueba
print(f'Conjunto de entrenamiento: {X_train.shape}, Conjunto de prueba: {X_test.shape}')

print (X_train.shape, y_train.shape)
print (X_test.shape, y_test.shape)


Conjunto de entrenamiento: (559, 9), Conjunto de prueba: (140, 9)
(559, 9) (559,)
(140, 9) (140,)


In [92]:
# Construccion del modelo de clasificación por regresión logística).
# Entrenamiento del modelo de clasificación por regresión logística
from sklearn.linear_model import LogisticRegression

# Creción del modelo de regresión logística
model = LogisticRegression(max_iter=500)  # Numero de iteraciones para converger

# Entrenamiento del modelo con los datos de entrenamiento
model.fit(X_train, y_train)





LogisticRegression(max_iter=500)

In [93]:
from sklearn.metrics import accuracy_score
# Se Prueba el modelo utilizando los datos de prueba.

predictions = model.predict(X_test)
predictions





array([2, 2, 2, 4, 2, 2, 2, 4, 4, 2, 2, 4, 2, 2, 2, 4, 2, 4, 4, 4, 4, 2,
       2, 2, 4, 2, 2, 4, 2, 4, 2, 2, 2, 4, 2, 2, 2, 4, 2, 4, 2, 2, 4, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 4, 2, 2, 2, 2, 4, 2, 2, 2, 2, 4, 2, 2, 2,
       4, 4, 2, 2, 4, 2, 4, 2, 2, 2, 2, 4, 2, 2, 2, 2, 2, 2, 2, 4, 2, 2,
       2, 2, 2, 2, 4, 2, 2, 4, 2, 2, 2, 2, 2, 2, 2, 4, 4, 2, 4, 2, 2, 4,
       4, 2, 2, 2, 2, 4, 2, 2, 2, 2, 2, 4, 2, 2, 4, 4, 2, 2, 4, 4, 2, 2,
       4, 2, 4, 2, 4, 2, 2, 2])

In [94]:
# Reportar la medida accuracy para el modelo construido.
# Cálculo del accuracy para evaluar el desempeño del modelo sobre los datos de prueba
from sklearn.metrics import accuracy_score

accuracy= accuracy_score(y_test, predictions)

# Mostrar la precisión
print(f'Precisión del modelo: {accuracy:.2f}')

Precisión del modelo: 0.96


**CONCLUSION**
Esto indica que el modelo tiene un 96% de precisión, lo que significa que, de todas las predicciones que hizo en el conjunto de prueba, el 96% fueron correctas.



# Fin del programa